tgui/packages/tgui/components/NanoMap.js

```diff
@@ -1,45 +1,179 @@
-        style={{
-          width: '512px',
-          height: '512px',
-        }}
-        onClick={onClick} />
-    </Box>
-  );
+const pauseEvent = e => {
+  if (e.stopPropagation) { e.stopPropagation(); }
+  if (e.preventDefault) { e.preventDefault(); }
+  e.cancelBubble = true;
+  e.returnValue = false;
+  return false;
 };
 
-const NanoMapMarker = props => {
+export class NanoMap extends Component {
+  constructor(props) {
+    super(props);
+
+    // Auto center based on window size
+    const Xcenter = (window.innerWidth / 2) - 256;
+    const Ycenter = (window.innerHeight / 2) - 256;
+
+    this.state = {
+      offsetX: 128,
+      offsetY: 48,
+      transform: 'none',
+      dragging: false,
+      originX: null,
+      originY: null,
+      zoom: 1,
+    };
+
+    // Dragging
+    this.handleDragStart = e => {
+      this.ref = e.target;
+      this.setState({
+        dragging: false,
+        originX: e.screenX,
+        originY: e.screenY,
+      });
+      document.addEventListener('mousemove', this.handleDragMove);
+      document.addEventListener('mouseup', this.handleDragEnd);
+      pauseEvent(e);
+    };
+
+    this.handleDragMove = e => {
+      this.setState(prevState => {
+        const state = { ...prevState };
+        const newOffsetX = e.screenX - state.originX;
+        const newOffsetY = e.screenY - state.originY;
+        if (prevState.dragging) {
+          state.offsetX += newOffsetX;
+          state.offsetY += newOffsetY;
+          state.originX = e.screenX;
+          state.originY = e.screenY;
+        } else {
+          state.dragging = true;
+        }
+        return state;
+      });
+      pauseEvent(e);
+    };
+
+    this.handleDragEnd = e => {
+      this.setState({
+        dragging: false,
+        originX: null,
+        originY: null,
+      });
+      document.removeEventListener('mousemove', this.handleDragMove);
+      document.removeEventListener('mouseup', this.handleDragEnd);
+      pauseEvent(e);
+    };
+
+    this.handleZoom = (_e, value) => {
+      this.setState(state => {
+        const newZoom = Math.min(Math.max(value, 1), 8);
+        let zoomDiff = (newZoom - state.zoom) * 1.5;
+        state.zoom = newZoom;
+        state.offsetX = state.offsetX - 262 * zoomDiff;
+        state.offsetY = state.offsetY - 256 * zoomDiff;
+        if (props.onZoom) {
+          props.onZoom(state.zoom);
+        }
+        return state;
+      });
+    };
+
+  }
+
+  render() {
+    const { config } = useBackend(this.context);
+    const { dragging, offsetX, offsetY, zoom = 1 } = this.state;
+    const { children } = this.props;
+
+    const mapUrl = config.map + "_nanomap_z1.png?" + Date.now();
+    const mapSize = (510 * zoom) + 'px';
+    const newStyle = {
+      width: mapSize,
+      height: mapSize,
+      "margin-top": offsetY + "px",
+      "margin-left": offsetX + "px",
+      "overflow": "hidden",
+      "position": "relative",
+      "background-image": "url(" + mapUrl + ")",
```

Comment:
Can you try using `image-rendering` here to preserve the pixelart-ness of the images?

Reply:
Tried, doesn't work.